apsuite/commissioning_scripts/meas_coupling_tune.py

```diff
@@ -17,201 +47,207 @@ class FitTunes(_SimulAnneal):
-        _plt.title(title)
-        _plt.grid()
-        _plt.show()
-
-    def _calc_init_parms(self):
-        """."""
-        parm = self._param
-        vyb = self._tune1[0], self._tune2[0]
-        vye = self._tune1[-1], self._tune2[-1]
-        dparm = parm[-1] - parm[0]
-        coeff1 = (min(vye) - max(vyb)) / dparm
-        offset1 = max(vyb) - coeff1 * parm[0]
-        coeff2 = (max(vye) - min(vyb)) / dparm
-        offset2 = min(vyb) - coeff2 * parm[0]
-        coupling = min(_np.abs(self._tune1 - self._tune2))
-        position = [coeff1, offset1, coeff2, offset2, coupling]
-        self.position = position
+        axi.plot(qcurr_interp, fittune1, color='tab:gray', label='fitting')
+        axi.plot(qcurr_interp, fittune2, color='tab:gray')
+        axi.legend()
+        axi.set_xlabel(xlabel)
+        axi.set_ylabel('Transverse Tunes')
+        if save:
+            if fname is None:
+                date_string = _time.strftime("%Y-%m-%d-%H:%M")
+                fname = 'coupling_fitting_{}.png'.format(date_string)
+            fig.savefig(fname, format='png', dpi=300)
+        return fig, axi
+
+    def _filter_data(self):
+        qcurr = _np.asarray(self.data['current'])
+        tune1, tune2 = self.data['tunes'].T
+        tune1 = _np.asarray(tune1)
+        tune2 = _np.asarray(tune2)
+
+        # sort tune1 > tune2 at each point
+        sel = tune1 <= tune2
+        if sel.any():
+            tune1[sel], tune2[sel] = tune2[sel], tune1[sel]
+
+        # remove nearly degenerate measurement points
+        dtunes = _np.abs(tune1 - tune2)
+        sel = _np.where(dtunes < self.params.coupling_resolution)[0]
+        qcurr = _np.delete(qcurr, sel)
+        tune1 = _np.delete(tune1, sel)
+        tune2 = _np.delete(tune2, sel)
+        return qcurr, tune1, tune2
+
+    def _calc_fitting_error(self):
+        # based on fitting error calculation of scipy.optimization.curve_fit
+        # do Moore-Penrose inverse discarding zero singular values.
+        fit_params = self.analysis['fitted_param']
+        _, smat, vhmat = _np.linalg.svd(
+            fit_params['jac'], full_matrices=False)
+        thre = _np.finfo(float).eps * max(fit_params['jac'].shape)
+        thre *= smat[0]
+        smat = smat[smat > thre]
+        vhmat = vhmat[:smat.size]
+        pcov = _np.dot(vhmat.T / (smat*smat), vhmat)
+
+        # multiply covariance matrix by residue 2-norm
+        ysize = len(fit_params['fun'])
+        cost = 2 * fit_params['cost']  # res.cost is half sum of squares!
+        popt = fit_params['x']
+        if ysize > popt.size:
+            # normalized by degrees of freedom
+            s_sq = cost / (ysize - popt.size)
+            pcov = pcov * s_sq
+        else:
+            pcov.fill(_np.nan)
+            print('# of fitting parameters larger than # of data points!')
+        return _np.sqrt(_np.diag(pcov))
+
+    # static methods
+    @staticmethod
+    def _get_normal_modes(params, curr):
```

Comment:
This method should be part of the API (without the leading `_`). I will use it in the `MeasCouplingWindow`